parser_library/src/processing/opencode_provider.h

```diff
@@ -26,20 +45,122 @@ enum class ainsert_destination
     front,
 };
 
-// interface for hiding parser implementation
-class opencode_provider : public statement_provider
+struct opencode_provider_options
+{
+    bool ictl_allowed;
+    int process_remaining;
+};
+
+enum class extract_next_logical_line_result
+{
+    failed,
+    normal,
+    ictl,
+    process,
+};
+
+// uses the parser implementation to produce statements in the opencode(-like) scenario
+class opencode_provider final : public diagnosable_impl, public statement_provider
 {
+    struct lines_to_remove
+    {
+        size_t ainsert_buffer;
+        size_t copy_files;
+        size_t preprocessor_buffer;
+        size_t current_text_lines;
+    };
+    lines_to_remove m_lines_to_remove = {};
+
+    std::string_view m_original_text;
+    size_t m_current_line = 0;
+
+    std::string_view m_next_line_text;
+
+    lexing::logical_line m_current_logical_line;
+    struct logical_line_origin
+    {
+        size_t begin_offset;
+        size_t end_offset;
+        size_t begin_line;
+        size_t end_line;
+        enum class source_type
+        {
+            none,
+            file,
+            preprocessor,
+            copy,
+            ainsert,
+        } source;
+    } m_current_logical_line_source;
+
+    std::deque<std::string> m_ainsert_buffer;
+    std::vector<std::string_view> m_copy_files;
+
+    std::vector<std::string> m_preprocessor_buffer;
+
+    std::unique_ptr<parsing::parser_holder> m_parser;
+    std::unique_ptr<parsing::parser_holder> m_lookahead_parser;
+    std::unique_ptr<parsing::parser_holder> m_operand_parser;
+
+    analyzing_context* m_ctx;
+    workspaces::parse_lib_provider* m_lib_provider;
+    processing::processing_state_listener* m_state_listener;
+    semantics::source_info_processor* m_src_proc;
+
+    opencode_provider_options m_opts;
+
+    parsing::parser_error_listener m_listener;
+
+    bool m_line_fed = false;
+
 public:
     // rewinds position in file
-    virtual void rewind_input(context::source_position pos) = 0;
-    virtual std::string aread() = 0;
-    virtual void ainsert(const std::string& rec, ainsert_destination dest) = 0;
+    void rewind_input(context::source_position pos);
+    std::string aread();
+    void ainsert(const std::string& rec, ainsert_destination dest);
+
+    opencode_provider(std::string_view text,
+        analyzing_context& ctx,
+        workspaces::parse_lib_provider& lib_provider,
+        processing::processing_state_listener& state_listener,
+        semantics::source_info_processor& src_proc,
+        const std::string& filename,
```

Comment:
I am not 100% about this, but context has method `opencode_file_name`, probably could replace this parameter.

Reply:
there appear to be cases when we are passing both the context and the filename as parameters and at the moment I am not 100% convinced that they always match.